### PR TITLE
HTBHF-2010 Created client to encode and send events to google analytics.

### PR DESCRIPTION
**Pull request by dwybourn**
- The client creates the payload in the format of a query string with url encoded values (as required by google analytics). 
- Payload is sent to GA and asserts that we get a 200 response back (200 is the only valid response from GA)

### CODE REVIEW

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    // google analytics expects the payload to be url encoded and in the format of a query string
+    private static final String URL_ENCODED_REPORT_PROPERTIES = "cd8=Bristol+West&cd9=NHS+Bristol%2C+North+Somerset+and+South+Gloucestershire";
+
+    static {
+        REPORT_PROPERTIES.put("cd8", "Bristol West");
+        REPORT_PROPERTIES.put("cd9", "NHS Bristol, North Somerset and South Gloucestershire");
+    }
+
+    @Value("${google-analytics.base-uri}")
+    private String baseUri;
```
Comment (YetAnotherMatt) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java]: might be worth a comment explaining that this is set to localhost in test/resources/application.yml

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        GoogleAnalyticsException exception = catchThrowableOfType(() -> googleAnalyticsClient.reportEvent(REPORT_PROPERTIES), GoogleAnalyticsException.class);
+
+        assertThat(exception.getMessage()).isEqualTo("Response code from google analytics was not as expected, received: NO_CONTENT");
+        verify(exactly(1), postRequestedFor(urlEqualTo(REPORT_ENDPOINT))
+                .withHeader("Content-Type", equalTo(TEXT_PLAIN_VALUE))
+                .withRequestBody(equalTo(URL_ENCODED_REPORT_PROPERTIES)));
+    }
+
+    @Test
+    void shouldThrowExceptionWhenGoogleAnalyticsReturnsAndErrorResponse() {
```
Comment (YetAnotherMatt) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java]: typo: should be `shouldThrowExceptionWhenGoogleAnalyticsReturnsAnErrorResponse` (currently has `...AndErrorResponse`)

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java (hunk 1 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java]: Not sure if that's necessary as we haven't done so for any of the other properties set to localhost.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                                 RestTemplate restTemplate) throws URISyntaxException {
+        this.googleAnalyticsUri = new URI(baseUri + REPORT_ENDPOINT);
+        this.restTemplate = restTemplate;
+    }
+
+    public void reportEvent(Map<String, String> reportProperties) {
+        String payload = createQueryStringWithUrlEncodedValues(reportProperties);
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(TEXT_PLAIN);
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: Would it be worthwhile adding a comment (possibly in the JavaDoc for this method) to say why you needed to set this after all the pain you went through on Friday morning?

Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: Added a comment above the line.

Comment (mikeSmith08) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: I'd also explain what Google Analytics needs in terms of a payload to try to explain why its done this way.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Test
+    void shouldThrowExceptionWhenGoogleAnalyticsReturnsAnErrorResponse() {
+        stubFor(post(urlEqualTo(REPORT_ENDPOINT)).withHeader("Content-Type", equalTo(TEXT_PLAIN_VALUE))
+                .willReturn((serverError())));
+
+        GoogleAnalyticsException exception = catchThrowableOfType(() -> googleAnalyticsClient.reportEvent(REPORT_PROPERTIES), GoogleAnalyticsException.class);
+
+        assertThat(exception.getMessage()).isEqualTo("Exception caught trying to call google analytics at: " + baseUri + REPORT_ENDPOINT);
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java]: Would be good to also assert that the exception cause has been stored in the caught exception too.

Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClientTest.java]: Added assertion of the cause.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.stream.Collectors;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.springframework.http.HttpMethod.POST;
+import static org.springframework.http.HttpStatus.OK;
+import static org.springframework.http.MediaType.TEXT_PLAIN;
+
+/**
+ * Service for interacting with google analytics.
+ */
```
Comment (YetAnotherMatt) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: I'd suggest:
```
Service for interacting with google analytics measurement protocol.
See https://developers.google.com/analytics/devguides/collection/protocol/v1/
```

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java (hunk 3 above)
Comment (YetAnotherMatt) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: I've suggested adding a link to the measurement protocol documentation, which might cover it

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java (hunk 5 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: Updated.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.springframework.http.HttpMethod.POST;
+import static org.springframework.http.HttpStatus.OK;
+import static org.springframework.http.MediaType.TEXT_PLAIN;
+
+/**
+ * Service for interacting with google analytics measurement protocol.
+ * See https://developers.google.com/analytics/devguides/collection/protocol/v1/
+ */
+@Service
```
Comment (YetAnotherMatt) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: Shouldn't this be a `@Component` rather than `@Service`?

Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: Changed to Component.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java (hunk 3 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/reporting/GoogleAnalyticsClient.java]: I've left in the comment and added a link in the javadoc for the class.

**Pull request merged**